lib/ivis_opengl/gfx_api_gl.cpp

```diff
@@ -0,0 +1,87 @@
+#include "gfx_api.h"
+
+#include <GL/glew.h>
+#include <algorithm>
+#include <cmath>
+
+namespace gl_api
+{
+	auto to_gl(const gfx_api::pixel_format& format)
+	{
+		switch (format)
+		{
+		case gfx_api::pixel_format::rgba:
+			return GL_RGBA;
+		case gfx_api::pixel_format::rgb:
+			return GL_RGB;
+		case gfx_api::pixel_format::compressed_rgb:
+			return GL_RGB_S3TC;
+		case gfx_api::pixel_format::compressed_rgba:
+			return GL_RGBA_S3TC;
+		}
+		throw;
+	}
+
+	struct texture : public gfx_api::texture
+	{
+	private:
+		friend struct context;
+		GLuint _id;
+
+		texture()
+		{
+			glGenTextures(1, &_id);
+		}
+
+		~texture()
+		{
+			glDeleteTextures(1, &_id);
+		}
+	public:
+		virtual void bind() override
+		{
+			glBindTexture(GL_TEXTURE_2D, _id);
+		}
+
+		virtual void upload(const size_t& mip_level, const size_t& offset_x, const size_t& offset_y, const size_t & width, const size_t & height, const gfx_api::pixel_format & buffer_format, const void * data) override
+		{
+			bind();
+			glTexSubImage2D(GL_TEXTURE_2D, mip_level, offset_x, offset_y, width, height, to_gl(buffer_format), GL_UNSIGNED_BYTE, data);
+		}
+
+		virtual unsigned id() override
+		{
+			return _id;
+		}
+
+		virtual void generate_mip_levels() override
+		{
+			glGenerateMipmap(GL_TEXTURE_2D);
+		}
+
+	};
+
+	struct context : public gfx_api::context
+	{
+		virtual std::unique_ptr<gfx_api::texture> create_texture(const size_t & width, const size_t & height, const gfx_api::pixel_format & internal_format, const std::string& filename) override
+		{
+			auto* new_texture = new texture();
+			new_texture->bind();
+			if (!filename.empty() && (GLEW_VERSION_4_3 || GLEW_KHR_debug))
+			{
+				glObjectLabel(GL_TEXTURE, new_texture->id(), -1, filename.c_str());
+			}
+			for (unsigned i = 0; i < floor(log(std::max(width, height))); ++i)
+			{
+				glTexImage2D(GL_TEXTURE_2D, i, to_gl(internal_format), width >> i, height >> i, 0, to_gl(internal_format), GL_UNSIGNED_BYTE, nullptr);
+			}

```

Comment:
I wonder if we should just take the step up to glTexStorage2D() here instead. It is only core in OpenGL 4.0, but should be in most drivers that don't support 4.0 too, I think.